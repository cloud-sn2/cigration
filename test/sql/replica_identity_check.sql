/** 
测试环境信息
CN：127.0.0.1
    WORKER1：worker_1_host/57637
    WORKER2：worker_2_host/57638
    WORKER3：worker_3_host/57639

worker_1_host、worker_2_host、worker_3_host：都是127.0.0.1
*/

set datestyle = ISO, YMD;
set search_path = cigration_regress_test, cigration, public;
set citus.shard_count = 8;
set client_min_messages to warning;

drop schema if exists cigration_regress_test cascade;
create schema cigration_regress_test;

--
-- 1. 无主键分片表的迁移(with_replica_identity_check=true)
--

-- 创建无主键的分片表

create table nopk_tb1(c1 int, c2 text);
select create_distributed_table('nopk_tb1','c1', colocate_with=>'none');

-- 插入数据
insert into nopk_tb1 select generate_series(1,10000), 'aaa';

-- 查看所有分片的初始分布
select nodename,
       nodeport,
	   logicalrelid,
	   shardminvalue,
	   shardmaxvalue
from pg_dist_shard_placement p
     join pg_dist_shard s on(p.shardid = s.shardid)
order by nodename,nodeport,logicalrelid,shardminvalue;

-- 创建并执行迁移任务
select jobid, taskid from cigration_create_move_node_job(:'worker_1_host', :worker_1_port, :'worker_3_host', :worker_3_port) limit 1 \gset

select cigration_run_shard_migration_job(:jobid, with_replica_identity_check=>true);

-- 检查迁移任务状态，期待迁移异常
select source_nodename,source_nodeport,target_nodename,target_nodeport,status,total_shard_count,error_message
from pg_citus_shard_migration where jobid=:jobid and taskid=:taskid;

--
-- 2. 无主键分片表的迁移(with_replica_identity_check=false)
--

-- 清理错误环境
select cigration_cleanup_error_env();

-- 使用默认with_replica_identity_check参数（true）继续执行迁移任务
select cigration_run_shard_migration_job(:jobid);

select source_nodename,source_nodeport,target_nodename,target_nodeport,status,total_shard_count,error_message
from pg_citus_shard_migration;

-- 查看迁移后所有分片的分布
select nodename,
       nodeport,
	   logicalrelid,
	   shardminvalue,
	   shardmaxvalue
from pg_dist_shard_placement p
     join pg_dist_shard s on(p.shardid = s.shardid)
order by nodename,nodeport,logicalrelid,shardminvalue;

-- 检查分片迁移后SQL执行正常
select count(*) from nopk_tb1;

-- DROP表
drop table nopk_tb1;

--
-- 3. 无主键有replica_identity分片表的迁移(with_replica_identity_check=true)
--

-- 创建无主键有replica_identity的分片表
create table replica_identity_tb1(c1 int not null, c2 text);
create unique index idx_replica_identity_tb1_c1 on replica_identity_tb1(c1);
alter table replica_identity_tb1 REPLICA IDENTITY USING INDEX idx_replica_identity_tb1_c1 ;

select create_distributed_table('replica_identity_tb1','c1', colocate_with=>'none');

-- 插入数据
insert into replica_identity_tb1 select generate_series(1,10000), 'aaa';

-- 查看所有分片的初始分布
select nodename,
       nodeport,
	   logicalrelid,
	   shardminvalue,
	   shardmaxvalue
from pg_dist_shard_placement p
     join pg_dist_shard s on(p.shardid = s.shardid)
order by nodename,nodeport,logicalrelid,shardminvalue;

-- 创建并执行迁移任务
select jobid, taskid from cigration_create_move_node_job(:'worker_1_host', :worker_1_port, :'worker_3_host', :worker_3_port) limit 1 \gset

select cigration_run_shard_migration_job(:jobid, with_replica_identity_check=>true);

select source_nodename,source_nodeport,target_nodename,target_nodeport,status,total_shard_count,error_message
from pg_citus_shard_migration;

-- 查看迁移后所有分片的分布
select nodename,
       nodeport,
	   logicalrelid,
	   shardminvalue,
	   shardmaxvalue
from pg_dist_shard_placement p
     join pg_dist_shard s on(p.shardid = s.shardid)
order by nodename,nodeport,logicalrelid,shardminvalue;

-- 检查分片迁移后SQL执行正常
select count(*) from replica_identity_tb1;

-- DROP表
drop table replica_identity_tb1;


-- 4. 清空垃圾站
--

select nodename,nodeport,count(*)
from cigration_get_recyclebin_metadata()
group by nodename,nodeport
order by nodename,nodeport;

select cigration_cleanup_recyclebin();

select nodename,nodeport,count(*)
from cigration_get_recyclebin_metadata()
group by nodename,nodeport
order by nodename,nodeport;


--
-- 5. 测试环境清理
--
select cigration_cleanup_recyclebin();
-- 确认pg_citus_shard_migration中的记录为空
select count(*) from pg_citus_shard_migration;

