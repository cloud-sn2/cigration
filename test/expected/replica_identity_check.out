/** 
测试环境信息
CN：127.0.0.1
    WORKER1：worker_1_host/57637
    WORKER2：worker_2_host/57638
    WORKER3：worker_3_host/57639

worker_1_host、worker_2_host、worker_3_host：都是127.0.0.1
*/
set datestyle = ISO, YMD;
set search_path = cigration_regress_test, cigration, public;
set citus.shard_count = 8;
set client_min_messages to warning;
drop schema if exists cigration_regress_test cascade;
create schema cigration_regress_test;
--
-- 1. 无主键分片表的迁移(with_replica_identity_check=true)
--
-- 创建无主键的分片表
create table nopk_tb1(c1 int, c2 text);
select create_distributed_table('nopk_tb1','c1', colocate_with=>'none');
 create_distributed_table 
--------------------------
 
(1 row)

-- 插入数据
insert into nopk_tb1 select generate_series(1,10000), 'aaa';
-- 查看所有分片的初始分布
select nodename,
       nodeport,
	   logicalrelid,
	   shardminvalue,
	   shardmaxvalue
from pg_dist_shard_placement p
     join pg_dist_shard s on(p.shardid = s.shardid)
order by nodename,nodeport,logicalrelid,shardminvalue;
 nodename  | nodeport | logicalrelid | shardminvalue | shardmaxvalue 
-----------+----------+--------------+---------------+---------------
 localhost |    57637 | nopk_tb1     | 1073741824    | 1610612735
 localhost |    57637 | nopk_tb1     | -2147483648   | -1610612737
 localhost |    57637 | nopk_tb1     | -536870912    | -1
 localhost |    57638 | nopk_tb1     | 0             | 536870911
 localhost |    57638 | nopk_tb1     | 1610612736    | 2147483647
 localhost |    57638 | nopk_tb1     | -1610612736   | -1073741825
 localhost |    57639 | nopk_tb1     | -1073741824   | -536870913
 localhost |    57639 | nopk_tb1     | 536870912     | 1073741823
(8 rows)

-- 创建并执行迁移任务
select jobid, taskid from cigration_create_move_node_job(:'worker_1_host', :worker_1_port, :'worker_3_host', :worker_3_port) limit 1 \gset
select cigration_run_shard_migration_job(:jobid, with_replica_identity_check=>true);
ERROR:  task 71 in job 11 start failed.check errors in the task_info table:cigration.pg_citus_shard_migration.
CONTEXT:  PL/pgSQL function cigration.cigration_start_shard_migration_task(integer,integer,interval,boolean) line 33 at RAISE
while executing query on dblink connection named "cigration_run_shard_migration_job"
SQL statement "SELECT * FROM dblink('cigration_run_shard_migration_job', 
                                                      $sql$SELECT cigration.cigration_start_shard_migration_task(11, 71, longtime_tx_threshold=>'@ 30 mins', with_replica_identity_check=>'t')$sql$
                                                      ) AS t(result_record text)"
PL/pgSQL function cigration_run_shard_migration_job(integer,integer[],integer,interval,boolean) line 102 at EXECUTE
-- 检查迁移任务状态，期待迁移异常
select source_nodename,source_nodeport,target_nodename,target_nodeport,status,total_shard_count,error_message
from pg_citus_shard_migration where jobid=:jobid and taskid=:taskid;
 source_nodename | source_nodeport | target_nodename | target_nodeport | status | total_shard_count |                            error_message                             
-----------------+-----------------+-----------------+-----------------+--------+-------------------+----------------------------------------------------------------------
 localhost       |           57637 | localhost       |           57639 | error  |                 1 | it does not has a replica identity or it does not has a primary key.
(1 row)

--
-- 2. 无主键分片表的迁移(with_replica_identity_check=false)
--
-- 清理错误环境
select cigration_cleanup_error_env();
 cigration_cleanup_error_env 
-----------------------------
 
(1 row)

-- 使用默认with_replica_identity_check参数（true）继续执行迁移任务
select cigration_run_shard_migration_job(:jobid);
 cigration_run_shard_migration_job 
-----------------------------------
 t
(1 row)

select source_nodename,source_nodeport,target_nodename,target_nodeport,status,total_shard_count,error_message
from pg_citus_shard_migration;
 source_nodename | source_nodeport | target_nodename | target_nodeport | status | total_shard_count | error_message 
-----------------+-----------------+-----------------+-----------------+--------+-------------------+---------------
(0 rows)

-- 查看迁移后所有分片的分布
select nodename,
       nodeport,
	   logicalrelid,
	   shardminvalue,
	   shardmaxvalue
from pg_dist_shard_placement p
     join pg_dist_shard s on(p.shardid = s.shardid)
order by nodename,nodeport,logicalrelid,shardminvalue;
 nodename  | nodeport | logicalrelid | shardminvalue | shardmaxvalue 
-----------+----------+--------------+---------------+---------------
 localhost |    57638 | nopk_tb1     | 0             | 536870911
 localhost |    57638 | nopk_tb1     | 1610612736    | 2147483647
 localhost |    57638 | nopk_tb1     | -1610612736   | -1073741825
 localhost |    57639 | nopk_tb1     | 1073741824    | 1610612735
 localhost |    57639 | nopk_tb1     | -1073741824   | -536870913
 localhost |    57639 | nopk_tb1     | -2147483648   | -1610612737
 localhost |    57639 | nopk_tb1     | 536870912     | 1073741823
 localhost |    57639 | nopk_tb1     | -536870912    | -1
(8 rows)

-- 检查分片迁移后SQL执行正常
select count(*) from nopk_tb1;
 count 
-------
 10000
(1 row)

-- DROP表
drop table nopk_tb1;
--
-- 3. 无主键有replica_identity分片表的迁移(with_replica_identity_check=true)
--
-- 创建无主键有replica_identity的分片表
create table replica_identity_tb1(c1 int not null, c2 text);
create unique index idx_replica_identity_tb1_c1 on replica_identity_tb1(c1);
alter table replica_identity_tb1 REPLICA IDENTITY USING INDEX idx_replica_identity_tb1_c1 ;
select create_distributed_table('replica_identity_tb1','c1', colocate_with=>'none');
 create_distributed_table 
--------------------------
 
(1 row)

-- 插入数据
insert into replica_identity_tb1 select generate_series(1,10000), 'aaa';
-- 查看所有分片的初始分布
select nodename,
       nodeport,
	   logicalrelid,
	   shardminvalue,
	   shardmaxvalue
from pg_dist_shard_placement p
     join pg_dist_shard s on(p.shardid = s.shardid)
order by nodename,nodeport,logicalrelid,shardminvalue;
 nodename  | nodeport |     logicalrelid     | shardminvalue | shardmaxvalue 
-----------+----------+----------------------+---------------+---------------
 localhost |    57637 | replica_identity_tb1 | 1073741824    | 1610612735
 localhost |    57637 | replica_identity_tb1 | -2147483648   | -1610612737
 localhost |    57637 | replica_identity_tb1 | -536870912    | -1
 localhost |    57638 | replica_identity_tb1 | 0             | 536870911
 localhost |    57638 | replica_identity_tb1 | 1610612736    | 2147483647
 localhost |    57638 | replica_identity_tb1 | -1610612736   | -1073741825
 localhost |    57639 | replica_identity_tb1 | -1073741824   | -536870913
 localhost |    57639 | replica_identity_tb1 | 536870912     | 1073741823
(8 rows)

-- 创建并执行迁移任务
select jobid, taskid from cigration_create_move_node_job(:'worker_1_host', :worker_1_port, :'worker_3_host', :worker_3_port) limit 1 \gset
select cigration_run_shard_migration_job(:jobid, with_replica_identity_check=>true);
 cigration_run_shard_migration_job 
-----------------------------------
 t
(1 row)

select source_nodename,source_nodeport,target_nodename,target_nodeport,status,total_shard_count,error_message
from pg_citus_shard_migration;
 source_nodename | source_nodeport | target_nodename | target_nodeport | status | total_shard_count | error_message 
-----------------+-----------------+-----------------+-----------------+--------+-------------------+---------------
(0 rows)

-- 查看迁移后所有分片的分布
select nodename,
       nodeport,
	   logicalrelid,
	   shardminvalue,
	   shardmaxvalue
from pg_dist_shard_placement p
     join pg_dist_shard s on(p.shardid = s.shardid)
order by nodename,nodeport,logicalrelid,shardminvalue;
 nodename  | nodeport |     logicalrelid     | shardminvalue | shardmaxvalue 
-----------+----------+----------------------+---------------+---------------
 localhost |    57638 | replica_identity_tb1 | 0             | 536870911
 localhost |    57638 | replica_identity_tb1 | 1610612736    | 2147483647
 localhost |    57638 | replica_identity_tb1 | -1610612736   | -1073741825
 localhost |    57639 | replica_identity_tb1 | 1073741824    | 1610612735
 localhost |    57639 | replica_identity_tb1 | -1073741824   | -536870913
 localhost |    57639 | replica_identity_tb1 | -2147483648   | -1610612737
 localhost |    57639 | replica_identity_tb1 | 536870912     | 1073741823
 localhost |    57639 | replica_identity_tb1 | -536870912    | -1
(8 rows)

-- 检查分片迁移后SQL执行正常
select count(*) from replica_identity_tb1;
 count 
-------
 10000
(1 row)

-- DROP表
drop table replica_identity_tb1;
-- 4. 清空垃圾站
--
select nodename,nodeport,count(*)
from cigration_get_recyclebin_metadata()
group by nodename,nodeport
order by nodename,nodeport;
 nodename  | nodeport | count 
-----------+----------+-------
 localhost |    57637 |     2
 localhost |    57638 |     2
 localhost |    57639 |     2
(3 rows)

select cigration_cleanup_recyclebin();
 cigration_cleanup_recyclebin 
------------------------------
 
(1 row)

select nodename,nodeport,count(*)
from cigration_get_recyclebin_metadata()
group by nodename,nodeport
order by nodename,nodeport;
 nodename | nodeport | count 
----------+----------+-------
(0 rows)

--
-- 5. 测试环境清理
--
select cigration_cleanup_recyclebin();
 cigration_cleanup_recyclebin 
------------------------------
 
(1 row)

-- 确认pg_citus_shard_migration中的记录为空
select count(*) from pg_citus_shard_migration;
 count 
-------
     0
(1 row)

